examples/07_end_to_end/earthquake.ipynb

```diff
@@ -0,0 +1,416 @@
+   "id": "20ebbd2b-2bc6-445f-a53b-920324e7ad5b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "catalog"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6efee42c-41d9-4f42-b6f9-1eaa1170c653",
+   "metadata": {},
+   "source": [
+    "This catalog's events are not ordered, but we can sort them easily:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "aad93bfe-b889-44dc-9f48-a9ba36fdf7b9",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def sorting(event):\n",
+    "    return event.origins[0].time\n",
+    "catalog.events.sort(key=sorting)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "dc6eeaeb-17ec-4656-9675-112ab5552317",
+   "metadata": {},
+   "source": [
+    "Now, we can calso use [cartopy](https://scitools.org.uk/cartopy/docs/latest/) to to plot all events in a world map:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "2786ced2-1e5c-4c69-84ab-e4a4c16c7547",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "catalog.plot();"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f0d17801-aadd-4c2c-b257-0b30de7e7b7d",
+   "metadata": {},
+   "source": [
+    "To see what these event objects are, we can look closer:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "692262d9-f217-4fa0-8140-2d5641f6aa75",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "event = catalog[0]\n",
+    "event"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "bd33d70d-da4b-41a6-bf72-01cc0bdcb875",
+   "metadata": {},
+   "source": [
+    "This feed lists different kinds of events, noted by `event.event_type`. We can also inspect location, time and magnitude information:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "81b272b6-ab8f-4853-a115-2a6fbc94a335",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "event.origins, event.magnitudes"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c94fd3cc-5377-498d-ab2c-5b528a87d93c",
+   "metadata": {},
+   "source": [
+    "This feed is updated every minute, meaning we get a historical dataset of all seismic events in the past 24h, but we also get a continually updated feed that adds new events as they are entered into this feed (all events will contain this `creation_time` information)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "390b601f-8d13-496f-a3ca-820b08485009",
+   "metadata": {},
+   "source": [
+    "## With CSP"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f3b61d0a-1f3c-4d26-937a-4c14f05f757d",
+   "metadata": {},
+   "source": [
+    "We can now use CSP to read the same data in either realtime or simulation modes, by building a `PushPullAdapter`. This adapter combines a realtime or `PushAdapter` and a historical or `PullAdapter` into a single implementation. This makes it easy to switch from realtime to historical mode by the `realtime` argument of `csp.run`."
```

Comment:
I would emphasize the point of PushPull is in order to seamless switch from historical to realtime *at runtime*
There really is no point in using PushPull adapter with realtime=False, I would remove that comment.  Note how we can start our engine in the past and into the future, and csp will seamless run on the historical data and switch to pushing the realtime data